enzyme/test/Integration/ReverseMode/stl_list.cpp

```diff
@@ -0,0 +1,103 @@
+// FIXME: -O0 fails reverse mode (wrong result) https://github.com/EnzymeAD/Enzyme/pull/2370#issuecomment-3046307237
+// RUN: %clang++ -std=c++11 -fno-exceptions -ffast-math -O1 %s -S -emit-llvm -o - | %opt - %OPloadEnzyme %enzyme -S | %lli -
+// RUN: %clang++ -std=c++11 -fno-exceptions -ffast-math -O2 %s -S -emit-llvm -o - | %opt - %OPloadEnzyme %enzyme -S | %lli -
+// RUN: %clang++ -std=c++11 -fno-exceptions -ffast-math -O3 %s -S -emit-llvm -o - | %opt - %OPloadEnzyme %enzyme -S | %lli -
+
+#include "../test_utils.h"
+
+#include <iostream>
+#include <list>
+
+
+struct S {
+    S(double r) : x(r) {};
+    double x = 0.0;
+};
+
+extern double __enzyme_fwddiff(void*, int, std::list<double>&, int, ...);
+extern double __enzyme_autodiff(void*, int, std::list<double>&, int, ...);
+extern double __enzyme_fwddiff(void*, int, std::list<S>&, int, ...);
+extern double __enzyme_autodiff(void*, int, std::list<S>&, int, ...);
+
+
+double test_iterate_list(std::list<double>& vals, double const & x) {
+    // iterate over list
+    double result = 0.0;
+    for (const auto& val : vals) {
+        result += val * val * x;
+    }
+    return result;
+}
+
+double test_modify_list(std::list<S> & vals, double const & x) {
+    // simplified function for comparison:
+    //return x*x;
+
+    vals.front().x = x;
```

Comment:
@wsmoses this test would be an important use case for ImpactX particle accelerator beamlines: working on an existing list of structs and modifying some of their attributes based on `x`, then tracking derivatives on how these structs are used.

This does not work even in forward mode yet, which makes me think that we might some special treatment for `std::list<T>::front()` (and maybe also iterators in general)?

Reply:
Solution: pass vals as `enzyme_dup`:
https://fwd.gymni.ch/cJlqgX